### ISSUE TITLE
Only the first match per line is highlighted

### ISSUE THREAD

**Issue by gnprice** [livegrep/livegrep#67]
E.g., this query:
https://livegrep.com/search/linux?q=rx_length_errors+file%3Artnetl
matches a line
```
	a->rx_length_errors = b->rx_length_errors;
```
but only the first `rx_length_errors` is highlighted, not the other one.

This is potentially pretty confusing if e.g. you're searching for places where some field (like `rx_length_errors`) is read from, and you don't see any -- worse, you see some but not all so you don't immediately realize anything is wrong -- because when you scan the results page for the matches, the one you're looking for isn't highlighted. At a quick test, GNU `grep`, `git grep`, `ack`, `ag, and `rg` all highlight both matches, so that's also what users will expect.

I think to fix this I'd probably extend the `SearchResult` type in `src/proto/livegrep.proto` and the corresponding `Result` type in `server/api/types.go` so that their respective `Bounds`/`bounds` members represented not one but a sequence of intervals within the line. Then I think it'll be pretty straightforward how each layer in turn adapts to provide/transmit/consume that data.

The alternative I'd consider is to instead send multiple `Result` / `SearchResult` messages per line when there are multiple matches. This feels less good to me partly because it's more overhead especially when there are a lot of matches on a line, and maybe more fundamentally because it opens up the possibility of getting a message for an arbitrary subset of the matches on a line, which feels off. @nelhage may have other thoughts or another idea entirely, though.

Thanks to @jhurwitz for pointing this out on our internal instance at Dropbox :)

**Comment by nelhage** [comment]
Yeah, agreed that extending the match type to have a range of bounds seems like a better solution, and probably not a huge change. It also seems easy to make backwards compatible by adding the array as a new field, if need be (But I don't think we'd even need to).

**Comment by xvandish** [comment]
Hello, just wanted to give you a heads up I'm working on this @nelhage ! [Overview of the work so far.](https://github.com/xvandish/livegrep/compare/my_main...xvandish:livegrep:fix/highlight-all-matches) And a screenshot - 
![Screen Shot 2022-08-12 at 1 31 27 PM](https://user-images.githubusercontent.com/23644345/184440088-0bef626d-f901-48f5-96e5-06e2e03a3d4d.png)


My idea so far and some random thoughts.

Once `try_match` has been called and we confirm a match, we should search the matching line and it's context lines for additional matches.
* For regex expressions, we can call `Match` in a loop, iterating by the length of the last match, if any. 
* For literal expressions, we can speed things up by using a literal search instead of RE2. I'm pretty much just copying ideas from Rusts regex crate and all the work that @BurntSushi has done. So far I've implemented RabinKarp and simple memchar literal matchers, aho-corasick and TwoWay search would be next (if necessary). Since we know where matches are, generally, we don't have the same performance requirements as grep/ripgrep, so string search past RabinKarp may not be necessary, or at least not for a first version
* I think the hardest part of this work is just going to be figuring out whether an input pattern is a literal or not. The webserver knows, and acts accordingly, and maybe we could pass that info along, but I think we should also play it safe and have the codesearch backend decide whether something is a literal or not.
    * RE2 doesn't expose whether something is a literal or not, unlike the rust crate, and simple members like `prefix_` are private.
    * I had a convo on the [mailing list](https://groups.google.com/g/re2-dev/c/0fR8MbBCYXg) and the recommendation was, if we want to depend on RE2, to walk the syntax tree. If not, figure out it out ourselves - (maybe I could just check for the presence of any of the regex special operators unescaped, and some parens/brackets balancing).
* oh, and once we know something is a literal, we still need to deal with case insensitive alternations, and allowing the literal matchers to match on any of those. 
* For literals, the Rust crate does something really cool, picking the most unique/rare bytes out of a prefix and searching for those as a way to filter down the time spent matching.
* We'll probably want a bounded cache of StringPieces we've searched - any advice for this portion? Since StringPieces are just windows into an existing thing, do they need to be turned into strings to be cached as a key? Or will their use in the cache prompt C++ to not call the destructor? For clarity, I'm thinking something like this

**Comment by BurntSushi** [comment]
Just a small note: if you're trying to follow in my footsteps, especially when it comes to literal optimizations, keep in mind that it's entirely a black art. I'm almost positive that the culmination of my work falls in a local optimum somewhere. So don't be afraid to do something different. I myself have plans to totally rethink my approach in the next year or so.

**Comment by xvandish** [comment]
> Just a small note: if you're trying to follow in my footsteps, especially when it comes to literal optimizations, keep in mind that it's entirely a black art. I'm almost positive that the culmination of my work falls in a local optimum somewhere. So don't be afraid to do something different. I myself have plans to totally rethink my approach in the next year or so.

Thanks for the advice @BurntSushi, I'll keep it in mind :). Oh, and thanks for ripgrep, I use it literally everyday, it's amazing!

**Comment by nelhage** [comment]
Is it necessary to do a literal optimization at all, here? We know we're only matching within the scope of a single line in this case, so I would expect it'd be fine to just call directly into RE2. Especially since I believe it has internal optimizations for literal prefixes, which I expect will get us most of the benefits.

Did you run any benchmarking or performance numbers that suggested this was particularly expensive? My instincts could be wrong here.

What approach did you take to evolving the protocol? That seems more sensitive to me and (somewhat) more expensive to make changes to in the future, since it requires coordinated rollouts of the client&server.

**Comment by xvandish** [comment]
> Is it necessary to do a literal optimization at all, here? We know we're only matching within the scope of a single line in this case, so I would expect it'd be fine to just call directly into RE2. Especially since I believe it has internal optimizations for literal prefixes, which I expect will get us most of the benefits.
> 


I don't think it's necessary, definitely not at first. I just rushed in and it seemed like a fun thing to do, no benchmarking that made me think literal optimizations were necessary - but like you said, as we're only doing single searches within a single line, we should be fine - and that's seems to be the case in the code I have that [does that](https://github.com/xvandish/livegrep/commit/b03817cf63df67d36d02eb4e053e146dbfd5021e). 

> What approach did you take to evolving the protocol? That seems more sensitive to me and (somewhat) more expensive to make changes to in the future, since it requires coordinated rollouts of the client&server.

Yeah, the approach I've taken so far I'm not sure is the one I think would be best. I basically put the new fields and type of fields and stuck them in `match_result` under a `v2` prefix. So
```c++
struct match_result {
  StringPiece line
  vector<StringPiece> context_before;
  vector<StringPiece>context_after;
   int matchleft, matchright;

  // new fields
  vector<context_line> context_before_v2;
  vector<context_line> context_after_v2;
  vector<match_bound> match_bounds_v2;
}
```
The good thing about this approach is the webserver and client js can do things conditionally based on the presence of the V2 fields. The bad thing is later getting rid of the V2 suffix and righting towards a clean structure seems a pain (we can provide multiple commits to let people slowly roll client&server, but eh...)

I think I'd be in favor of just reworking `match_result` completely in a breaking way. That way, the upgrade wouldn't involve "flags" in the webserver and client js, and wouldn't involve more work on my end in providing multiple commits that people can run off of. Here's a diff: 
```diff
diff --git a/src/codesearch.h b/src/codesearch.h
index cb32e65..424a23e 100644
--- a/src/codesearch.h
+++ b/src/codesearch.h
@@ -92,13 +92,23 @@ struct index_info {
     vector<indexed_tree> trees;
 };
 
+struct match_bound {
+    int matchleft;
+    int matchright;
+};
+
+// empty bounds mean no match for q.line_pat on this line
+struct result_line {
+    StringPiece line;
+    vector<match_bound> match_bounds;
+};
+
 struct match_result {
     indexed_file *file;
     int lno;
-    vector<StringPiece> context_before;
-    vector<StringPiece> context_after;
-    StringPiece line;
-    int matchleft, matchright;
+    result_line line;
+    vector<result_line> context_before;
+    vector<result_line> context_after;
 };
 
 struct file_result {
```

**Comment by xvandish** [comment]
@juliannadeau-stripe any idea how "alright" people would be with breaking changes to the codesearch API at Stripe (like how many people use/make tools that query the `/api/v1/` api directly, other than the `lg` cli? 

I'm trying to weigh the cost of stuffing the new response structure into a v2 api, and making the requisite changes everywhere (there'd be a lot), vs just breaking the current api, which requires much less code.

**Comment by xvandish** [comment]
Oh, tagging @LINKIWI as well, I know you were using/created Webgrep, which if I remember right has a Node server call out to the c++ backend directly - what do you think about the above? For you I guess it would the questions is, would you be alright with breaking proto changes or would you prefer a `SearchResultV2` kind of thing along with a `SearchV2` rpc call?

**Comment by xvandish** [comment]
@asutherland, looks like you've been working on mozsearch recently, if you have the time could you weigh in on the convo above about changes to the Search RPC/interface? tks!

**Comment by LINKIWI** [comment]
> Oh, tagging @LINKIWI as well, I know you were using/created Webgrep, which if I remember right has a Node server call out to the c++ backend directly - what do you think about the above? For you I guess it would the questions is, would you be alright with breaking proto changes or would you prefer a `SearchResultV2` kind of thing along with a `SearchV2` rpc call?

Yes, it's a Node server that talks to livegrep over gRPC. That said though, I've since left Pinterest so I can't speak to its current usage, and I believe they have migrated to another code search solution anyway.

More generally though, maintaining API backwards compatibility is good practice that should, in theory at least, be easier to do with Protobuf. However, all APIs evolve so if you need to make a breaking change to support new functionality that's fine too. Nelson and/or other maintainers should probably be the ones to make that call.

**Comment by juliannadeau-stripe** [comment]
@xvandish about 0.1% of our traffic to that endpoint comes from Golang bindings and Curl, so I'm sure someone has it set up, but we haven't really operationalized around livegrep

**Comment by nelhage** [comment]
Yeah, I'm a bit torn; on the one hand it wouldn't be too hard to do something like the `_v2` solution that maintains compatibility. But also I expect most deployments are pretty tightly-coupled between client and server and upgrade in sync, so it may not be worth the trouble. I'd probably break towards just evolving forward in an incompatible way, so that the migration is done with.

I think the other question to think about is whether it's worthwhile to send match regions for every context line, or just the "result" line. The current webui will "merge" results that are close enough into one UI element, which has the effect that, in most cases, you'll get highlighting even for additional matches in the context, anyways.

**Comment by asutherland** [comment]
> @asutherland, looks like you've been working on mozsearch recently, if you have the time could you weigh in on the convo above about changes to the Search RPC/interface? tks!

Thanks for the ping!  I also agree that a breaking change seems most pragmatic (and most optimal for project health!).  mozsearch pins the version of livegrep we use so it's not like we would accidentally upgrade and break.  (And we have integration tests that would catch things quite quickly when we upgrade, etc.)

Re: matches for context, mozsearch does a similar coalescing to what @nelhage says livegrep does here in terms of eliding context lines in favor of match lines where those hunks overlap, so it doesn't seem like match results for context lines would be necessary/helpful.

**Comment by xvandish** [comment]
Thanks all for taking the time to respond! I've put up PR #349 based on the feedback in your responses, if any q's or feedback at all let me know - happy to discuss.

> Yes, it's a Node server that talks to livegrep over gRPC. That said though, I've since left Pinterest so I can't speak to its current usage, and I believe they have migrated to another code search solution anyway.

Sounds good, thanks!

> @xvandish about 0.1% of our traffic to that endpoint comes from Golang bindings and Curl, so I'm sure someone has it set up, but we haven't really operationalized around livegrep.
> 
> Digging in, I likely have about a dozen users (who would be "power users") I'd have to notify, so as long as there's a small write up I can give them - I think we're 🆒 😄

Perfect! if the PR I have up (#349) goes through, I'll get something written up for you.



> Yeah, I'm a bit torn; on the one hand it wouldn't be too hard to do something like the `_v2` solution that maintains compatibility. But also I expect most deployments are pretty tightly-coupled between client and server and upgrade in sync, so it may not be worth the trouble. I'd probably break towards just evolving forward in an incompatible way, so that the migration is done with.

👍. This is the direction I've gone in in the PR.

> I think the other question to think about is whether it's worthwhile to send match regions for every context line, or just the "result" line. The current webui will "merge" results that are close enough into one UI element, which has the effect that, in most cases, you'll get highlighting even for additional matches in the context, anyways.

Ah right, good catch. The reason I thought this was more necessary was actually a regression I introduced in my fork that made it so context lines without bounds "won" over match lines with bounds sometimes, so there were lots of bare/non-highlighted. Once I fixed that up all was good, and like you said, most of the time we get highlighting for context matches. 

> Thanks for the ping! I also agree that a breaking change seems most pragmatic (and most optimal for project health!). mozsearch pins the version of livegrep we use so it's not like we would accidentally upgrade and break. (And we have integration tests that would catch things quite quickly when we upgrade, etc.)

Np and perfect! This is the direction I've gone in in the PR.

> Re: matches for context, mozsearch does a similar coalescing to what @nelhage says livegrep does here in terms of eliding context lines in favor of match lines where those hunks overlap, so it doesn't seem like match results for context lines would be necessary/helpful.

 Sounds good, see above for why I thought searching context was more necessary than it really was - the PR only extends match lines, not context lines now thanks to you guy's feedback/reminder.

Once again, tks all for taking the time, appreciate it.
